app/Http/Controllers/Forum/TopicsController.php

```diff
@@ -272,63 +274,83 @@ public function show($id)
 
         priv_check('ForumView', $topic->forum)->ensureCan();
 
-        $posts = $topic->posts()->showDeleted($showDeleted);
-
-        if ($postStartId === 'unread') {
-            $postStartId = Post::lastUnreadByUser($topic, $currentUser);
-        } else {
-            $postStartId = get_int($postStartId);
-        }
+        if ($params['cursor'] === null) {
+            if ($params['start'] === 'unread') {
+                $params['start'] = Post::lastUnreadByUser($topic, $currentUser);
+            } else {
+                $params['start'] = get_int($params['start']);
+            }
 
-        if ($nthPost !== null) {
-            $post = $topic->nthPost($nthPost);
-            if ($post) {
-                $postStartId = $post->post_id;
+            if ($params['n'] !== null) {
+                $post = $topic->nthPost($params['n']);
+                if ($post !== null) {
+                    $params['cursor'] = ['id' => $post->getKey() - 1];
+                    $params['sort'] = 'id_asc';
+                }
+            } elseif ($params['start'] !== null) {
+                $params['cursor'] = ['id' => $params['start'] - 1];
+                $params['sort'] = 'id_asc';
+            } elseif ($params['end'] !== null) {
+                $params['cursor'] = ['id' => $params['end'] + 1];
+                $params['sort'] = 'id_desc';
```

Comment:
The 404 guards near the start of the function should probably be done even earlier and the `$params` building split out of this long function.